File: 02-programacion-basica/15-integracion/java/soluciones/RetoRegistroEstudiantes.java
import java.util.ArrayList;
import java.util.List;

public class RetoRegistroEstudiantes {
    public static void main(String[] args) {
// Las listas paralelas mantienen relacionados nombre y nota por posición.
        List<String> nombres = new ArrayList<>();
        List<Integer> notas = new ArrayList<>();

        registrar(nombres, notas, "Ana", 18);
        registrar(nombres, notas, "Luis", 14);
        registrar(nombres, notas, "María", 11);

        listar(nombres, notas);
        System.out.printf("Promedio: %.2f%n", promedio(notas));
    }

    static void registrar(List<String> nombres, List<Integer> notas, String nombre, int nota) {
        if (nombre == null || nombre.isBlank()) throw new IllegalArgumentException("Nombre vacío");
        if (nota < 0 || nota > 20) throw new IllegalArgumentException("Nota fuera de rango");
        nombres.add(nombre.trim());
        notas.add(nota);
    }

    static void listar(List<String> nombres, List<Integer> notas) {
        for (int i = 0; i < nombres.size(); i++) {
            System.out.printf("%s -> %d%n", nombres.get(i), notas.get(i));
        }
    }

    static double promedio(List<Integer> notas) {
        if (notas.isEmpty()) return 0;
        int suma = 0;
        for (int nota : notas) suma += nota;
        return (double) suma / notas.size();
    }
}

File: 02-programacion-basica/15-integracion/scala/soluciones/RetoRegistroEstudiantes.scala
def registrar(nombres: List[String], notas: List[Int], nombre: String, nota: Int): (List[String], List[Int]) =
  require(nombre.trim.nonEmpty, "Nombre vacío")
  require(nota >= 0 && nota <= 20, "Nota fuera de rango")
  (nombres :+ nombre.trim, notas :+ nota)

def promedio(notas: List[Int]): Double = if notas.isEmpty then 0.0 else notas.sum.toDouble / notas.size

@main def retoRegistroEstudiantes(): Unit =
  var nombres = List.empty[String]
  var notas = List.empty[Int]

  val r1 = registrar(nombres, notas, "Ana", 18); nombres = r1._1; notas = r1._2
  val r2 = registrar(nombres, notas, "Luis", 14); nombres = r2._1; notas = r2._2

  // zip mantiene relacionados los datos almacenados en las dos listas.
  nombres.zip(notas).foreach { case (nombre, nota) => println(s"$nombre -> $nota") }
  println(f"Promedio: ${promedio(notas)}%.2f")
